### PR TITLE
Added: Train custom Page-based Categorization AI

### PR DESCRIPTION
**Pull request by da-konfuzio**
This introduces a trainable Categorization AI composed of various Image Models and Text Models to predict per Page categorization.

Refer to the following documentation:
- [Python API docs - Categorization AI](https://dev.konfuzio.com/sdk/sourcecode.html#categorization-ai)
- [Tutorials - Model-based Categorization AI](https://dev.konfuzio.com/sdk/tutorials.html#model-based-categorization-ai)
- [Tutorials - Available Categorization Models](https://dev.konfuzio.com/sdk/tutorials.html#id8)
- [Tutorials - Categorization AI Overview Diagram](https://dev.konfuzio.com/sdk/tutorials.html#categorization-ai-overview-diagram)

### CODE REVIEW

File: konfuzio_sdk/trainer/tokenization.py [hunk 1]
```diff
@@ -0,0 +1,412 @@
...
+
+            # get examples to train phrase matcher
+            examples = train_dataset[label]
+
+            # build label-attr-phrase-matcher
+            phrase_matcher.add(label, list(spacy_model.tokenizer.pipe(examples)))
+
+        return phrase_matcher
+
+    def get_entities(self, text: str) -> List[Dict[str, Union[str, int]]]:
```
Comment (atraining) [inline konfuzio_sdk/trainer/tokenization.py:342]: why not working with document.spans

File: konfuzio_sdk/trainer/tokenization.py [hunk 2]
```diff
@@ -0,0 +1,412 @@
...
+    Returns a spacy language model.
+    """
+    try:
+        spacy_model = spacy.load(spacy_model_name)
+    except IOError:
+        raise IOError(f'Model not found, please install it with `python -m spacy download {spacy_model_name}`')
+    return spacy_model
+
+
+def get_tokenizer(tokenizer_name: str, project: Project, *args, **kwargs) -> Tokenizer:
```
Comment (atraining) [inline konfuzio_sdk/trainer/tokenization.py:398]: why do we need a wrapper if we can selelct the Class directly?

File: konfuzio_sdk/trainer/document_categorization.py [hunk 3]
```diff
@@ -114,3 +170,1841 @@ def categorize(self, document: Document, recategorize: bool = False, inplace: bo
...
+        """Forward pass."""
+        text = input['text']
+        # text = [batch size, seq len]
+        embeddings = self.dropout(self.embedding(text))
+        # embeddings = [batch size, seq len, emb dim]
+        embeddings = embeddings.permute(1, 0, 2)
+        # embeddings = [seq len, batch size, emb dim]
+        text_features, _ = self.lstm(embeddings)
+        # text_features = [seq len, batch size, hid dim * n directions]
+        text_features = text_features.permute(1, 0, 2)
```
[the remark this answers is not in the document]
Comment (da-konfuzio) [reply konfuzio_sdk/trainer/document_categorization.py:324]: I upgraded torch related dependencies to the latest version and this is no longer necessary, because the `MultiheadAttention` can be initialized with `batch_first=True` from `torch>=1.9.0`. The meaning was that it was expecting the input with shape `[seq len, batch, emb dim]` whereas the embeddings had shape `[batch, seq len, emb dim]`

https://pytorch.org/docs/stable/generated/torch.nn.MultiheadAttention.html#torch.nn.MultiheadAttention.forward

**Review by zypriafl — APPROVED** [review]
